#456921 [C#テキスト]4章演習問題

kurosu-being opened the pull request:
第4章　基本イディオム

コーディングレベルでよく利用されるイディオムを学習しました。
条件演算子の使用方法について学習しました。

kurosu-being committed:
#456921 [C#テキスト]4章演習問題

Practice4-1-1を実装しました。

kurosu-being committed:
#456921 [C#テキスト]4章演習問題

Practice4-1-2を実装しました。

kurosu-being committed:
#456921 [C#テキスト]4章演習問題

Practice4-1-3を実装しました。

kurosu-being committed:
#456921 [C#テキスト]4章演習問題

Practice4-1-4を実装しました。

kurosu-being committed:
#456921 [C#テキスト]4章演習問題

Practice4-2-1を実装しました。

kurosu-being committed:
#456921 [C#テキスト]4章演習問題

Practice4-2-1において、配列の命名を変更しました。
Practice4-2-2を実装しました。

kurosu-being committed:
#456921 [C#テキスト]4章演習問題

Practice4-2-3を実装しました。

kurosu-being committed:
#456921 [C#テキスト]4章演習問題

Practice4-2-5を実装しました。

kurosu-being committed:
#456921 [C#テキスト]4章演習問題

Practice4-2-5を実装しました。

kurosu-being committed:
#456921 [C#テキスト]4章演習問題

FindFirst21YearMonthメソッドについて、ヘッダーコメントを追加しました。

kurosu-being committed:
#456921 [C#テキスト]4章演習問題

AddOneMonthメソッドを問題文に合わせ、新しくインスタンスを作る形に変更しました。

kurosu-being committed:
#456921 [C#テキスト]4章演習問題

Practice4-1-3のコメントを修正しました。
プロパティで21世紀かのプロパティを読み取り専用に変更しました。

kurosu-being committed:
#456921 [C#テキスト]4章演習問題

不要な空行を削除しました。

k-miyahara-being commented on 章末問題4/Practice4-1-4/YearMonth.cs:
```diff
@@ -0,0 +1,53 @@
+        /// <returns>指定書式に変更した年月</returns>
+        public override string ToString() {
+            return ($"{Year}年{Month}月 ");
+        }
```
簡略化できます。
プロパティには`this`を付けておきましょう。
また、月の後に不要なスペースが入っています。

kurosu-being replied:
ご質問ありがとうございます！
`public override string ToString() => $"{this.Year}年{this.Month}月";`
に変更し、簡略化してthisを付ける形にいたしました。
中身の処理が簡単な時など、特に式本体を記述する方法に簡略化出来ないか試したいと思います。

k-miyahara-being commented on 章末問題4/Practice4-1-4/Program4-1-4.cs:
```diff
@@ -0,0 +1,10 @@
+    class Program {
+        //Practice4-1-4 ToStringメソッドをオーバーライドしてください。結果は"2017年8月"といった形式にしてください。
+        static void Main(string[] args) {
+            var wYearMonth = new YearMonth (2000, 3);
```
`new YearMonth (2000, 3)`
　　　　　 　↑のスペースは不要ですね。

kurosu-being replied:
ご指摘ありがとうございます！
不要なスペースを削除いたしました。

yuki-tanaka-being commented on 章末問題4/Practice4-1-2/YearMonth.cs:
```diff
@@ -0,0 +1,33 @@
+        /// </summary>
+        public bool Is21stCentury {
+            get {
+                return Year >= 2001 && Year <= 2100;
```
範囲の条件を記載するときは、`2001 <= Year && Year <= 2100` と記載すると可読性がよくなります。
リーダブルコードにも出てくるので、書籍を通して可読性について学んでみてください。

---

好みもありますが、省略系で書くとよりシンプルになります。
`public bool Is21stCentury => Year >= 2001 <= Year && Year <= 2100;`
特にGetオンリーのプロパティで1文なら、一行で書く方が可読性もよいかなと個人的には思います。

kurosu-being replied:
ご質問ありがとうございます！可読性を考慮し、式本体で書く省略形を適用して
`public bool Is21stCentury => 2001 <= Year && Year <= 2100;`
に修正しました。可読性についてリーダブルコードで勉強したいと思います。

yuki-tanaka-being commented on 章末問題4/Practice4-2-1/Program4-2-1.cs:
```diff
@@ -0,0 +1,14 @@
+            var wDay4 = new YearMonth(2023, 10);
+            var wDay5 = new YearMonth(2045, 7);
+            var wYearMonths = new YearMonth[] {wDay1, wDay2, wDay3, wDay4, wDay5,};
+        }
```
wDay1~5 までを変数に取る必要がないのなら直接コレクションの初期化子ないで new するほうがおすすめです。
変数の分だけ状態管理を意識する必要があります。
そのため、変数はできるだけ少ない方が、コードを読む際に頭の中で考えないといけないことが減らせて良いです。

可読性を考えるなら、改行して記載すると読みやすいです。
```
            var wYearMonths = new YearMonth[] {
                new YearMonth(2000, 3),
                new YearMonth(2010, 12),
                new YearMonth(2005, 2),
                new YearMonth(2023, 10),
                new YearMonth(2045, 7)
            };
```

kurosu-being replied:
ご指摘ありがとうございます！
ご提示いただいた変数を多用しない書き方に修正いたしました。
コレクションの初期化内で直接newをする、という考えが定着しておりませんでした。
以後変数をnewするときは特に気を付けて検討したいと思います。

yuki-tanaka-being commented on 章末問題4/Practice4-2-3/YearMonth.cs:
```diff
@@ -0,0 +1,66 @@
+        /// </summary>
+        /// <param name="wYearMonths">年月</param>
+        /// <returns>21世紀の年月</returns>
+        public static YearMonth FindFirst21YearMonth(YearMonth[] wYearMonths) {
```
メソッド名の Find は無くても意味が通じるかなと思いました。
`First21st(...)` でも十分かなと思います。
メソッド名は、その処理を実装しているクラス名とメソッドの引数と戻り値のシグネチャも込みで、ベストな命名を考えるとよいです。
今回の場合は、配列を受け取って配列の要素の方を返していることから、First21st だけで、配列の中から先頭の21st を返すと読み取れます。
ただし、削りすぎて意味が分からないよりは、まだ長い方がよいかなとは思います。

kurosu-being replied:
ご指摘ありがとうございます。
ご提案して頂いた、`First21st(...)`に命名を変更しました。
「メソッド名は『何をするか』で表す」という考えで固定されていて、根本の「一目で分かりやすいか」という問題を見逃していました。

> メソッド名は、その処理を実装しているクラス名とメソッドの引数と戻り値のシグネチャも込みで、ベストな命名を考えるとよいです。

そのメソッド、変数が何をしているのか端的に表せるよう「全体の中でどのような役割なのか」という観点を持ちたいと思います。

yuki-tanaka-being commented on 章末問題4/Practice4-2-3/YearMonth.cs:
```diff
@@ -0,0 +1,66 @@
+        /// </summary>
+        /// <param name="wYearMonths">年月</param>
+        /// <returns>21世紀の年月</returns>
+        public static YearMonth FindFirst21YearMonth(YearMonth[] wYearMonths) {
```
問題に指定がないのであれば、引数は最も広くとれる型を指定するのがよいです。
今回の場合は配列ではなく、`IEnumerable<YearMonth>` になります。
理由は、その方が汎用性が高まり、再利用しやすくなるからです。
最初から考えるのは大変なので、一旦実装してから、メソッド内で引数を使っている場所をみて、型を緩くできるなら書き換えるというリファクタリングのフェーズを挟むとよいです。
今回なら、foreach 構文でしか利用していないので、fereach 構文を利用するのに最低限必要な `IEnumerable<T>` が最も広く受けられる型となります。

余談ですが、演習問題ではなくプロダクトコードでも結構見落としがちです。
特に、処理を呼び出し側ありきで実装するとこうなります。
今回も、配列を定義して、その中からという誘導がありましたが、別にメソッドの引数の型を配列にしてくださいと指示があるわけではありません。
それなのに、呼び出し側が配列を渡すことが分かっているので、自然とメソッドの引数も配列にしたと思います。
本来メソッドは、呼び出し側に依存しないように作るのが大切です。
その処理の目的に対して、適切なイン（引数）とアウト（戻り値）を考えることが大切です。

kurosu-being replied:
ご指摘ありがとうございます。
インターフェイスを実装し、引数の型を配列より広くとれるように修正いたしました。
インターフェイスを適用する場合についてまだ理解が浅く、また、IEnumerable<T>がどんなメソッドやプロパティを持っているかなどの知識が少ないことにも気づきました。
まだ分かっていない箇所も多いですが、リファクタリングの際にインターフェイスが導入できないか考えることで理解を進めていこうと思います。

yuki-tanaka-being replied:
>インターフェイスを実装し、

用語の注意点。
一般的に「インターフェースを実装」とは、ある型に対してインターフェースを実装することを指します。
ここでの「実装」は抽象的な概念ではなく、「○○クラスがXXクラスを継承」のように具体的な状態を指します。

---

`IEnumerable<T>` に関しては、GoF本に出てくるデザインパターンの一つであるイテレーターパターンを学習すると理解が早まると思います。

以下は読み飛ばしてもらっても構いません。
イテレーターパターンを一言で説明すると、集約したオブジェクトを列挙する手段を提供するデザインパターンです。
C#の場合、イテレーターパターンは言語仕様として組み込まれています。
（モダンな言語ではほとんど標準で組み込まれています）
foreach で配列のループが回せるのはなぜ？同じく `List<T>` のループが回せるのはなぜ？
最初は、そういうものだと丸暗記で問題ないです。
原理を理解しようと一歩踏み込むと、イテレーターパターンで実現されていることが分かります。
具体的には、`IEnumerable` または `IEnumerable<T>` を実装する型であれば、foreach でループを回すことができます。
実際の型が配列だろうと `List<T>` だろうと `Dictionary<TKey, TValue>` だろうと、複数の要素を束ねる概念に対して、同様の手段でループを回せるようになります。
興味があれば、何も継承していない自作のクラスに `IEnumerable<T>` を実装してみると面白いと思います。
（yieldキーワードを使うと楽に実装できますが、イテレーターパターンを学ぶのであれば `IEnumerator<T>` を使う方法に挑んでみてほしいです）

と、これをすべて理解するのは大変だと思いますので、初学習者のうちは、コレクション系のクラスはすべて `IEnumerable<T>` が実装されていて、ループを回したいだけなら、`IEnumerable<T>` の型で問題ない。くらいを覚えておくだけでよいと思います。
※1点注意点として `IEnumerable<T>` は遅延評価が発生する場合があるため、実際に用いる場合はその辺りの理解度も必要になります。

yuki-tanaka-being commented on 章末問題4/Practice4-2-5/Program4-2-5.cs:
問題の意図と違う実装内容になっていると思います。
元の YearMonth オブジェクトを書き換えることは期待結果ではないと思います。
問題の意図は、4_1_3 で実装したメソッドを利用して、新たな YearMonth オブジェクトの配列を構築するだと思います。
> 配列に格納されているすべてのYearMonthの1ヶ月後を求め、その結果を新たな配列に入れてください。その後、その配列の要素の内容（年月）を順に表示してください。

今の実装は元の配列の要素の値を1ヶ月後に書き換えて、元の配列を年月の順にソートしたものを新たな配列に入れて、その配列を表示しています。

4_1_3 の `AddOneMonth()` と LINQ の `Select` を使って実装してみてください。

kurosu-being replied:
ご指摘ありがとうございます！
AddOneMonth() と LINQ の Select を使って、元の YearMonth オブジェクトを書き換えずに修正しました。

元のAddOneMonthメソッドを使ってコードでコーディングしていた際、
AddOneMonthメソッドを呼び出しても一か月後が表示されない問題を解消する為にAddOneMonthメソッド内部を書き直していました。その結果、本来の要求を歪めて解釈してしまったようです。
（本質的な原因はメソッドを呼ぶ際に新たな変数に変更後の配列を格納していないことでした。）

yuki-tanaka-being replied:
>（本質的な原因はメソッドを呼ぶ際に新たな変数に変更後の配列を格納していないことでした。）

ですね。ちゃんと問題点が理解できていて素晴らしいです。
これも人によってはあるあるです。
うまく行かないときに取る行動として、変な方向に突き進んでしまうパターン。
こういう時のおすすめは、他の人に話をするかなと思います。
答えを教えてもらうとかではなく、単に話を聞いてもらうだけでも、自分の中のモヤモヤや疑問点が整理できて、間違いに気づいたりしやすいです。

---

期待する動作としては問題ないですし、特に制約がなければ私も同じように実装するのですが、問題文の指示とはちょっとだけ反れているかもしれません。
問題文の「配列」が配列オブジェクトを指すのであれば、Select の結果を ToArray で一度変数にとり、その変数に対して OrderBy をしたものを直接 fereach で回すになるかと思います。
※指示がなければそんなことする必要性を全く感じませんが。。。

yuki-tanaka-being commented on 章末問題4/Practice4-2-5/Program4-2-5.cs:
```diff
@@ -0,0 +1,23 @@
+            for (int i = 0; i < wYearMonths.Length; i++) {
+                wYearMonths[i].AddOneMonth();
+            }
+            var wNextYearMonths = wYearMonths.OrderBy(x => new DateTime(x.Year, x.Month, 1)).ToArray();
```
ソートの実装なかなかおもしろいですね。
発想が柔軟でよいとなと思いました。
これでも実現できますが、ソートの度に無駄なインスタンスを生成するので効率が悪いです。

ソートの条件がそのオブジェクト自体を指すものであるなら、オブジェクト自身が実装を持つのがカプセル化の概念としてよいです。
一般的にオブジェクト自身にソート条件を実装する際は、[`IComparable<T>` ](https://learn.microsoft.com/ja-jp/dotnet/api/system.icomparable-1?view=netframework-4.7.2)インターフェースを実装します。
実装するメソッドは `int CompareTo(T other);` 1個です。
戻り値は 1 or 0 or -1 を返します。詳細は[リファレンス](https://learn.microsoft.com/ja-jp/dotnet/api/system.icomparable-1.compareto?view=netframework-4.7.2#system-icomparable-1-compareto(-0))参照。
![image](https://github.com/kurosu-being/TextC/assets/78195392/93cb2300-2a52-4425-a87f-b26a0196a39a)

今回だと、YearMonthクラスに、`IComparable<YearMonth>` を実装します。
例えば以下のような感じです。
```
        public int CompareTo(YearMonth other) =>
            this.Year > other.Year ? 1
            : this.Year < other.Year ? -1
            : this.Month > other.Month ? 1
            : this.Month < other.Month ? -1
            : 0;
```
その上で、呼び出し側は `wYearMonths.OrderBy(x => x)` とすれば、YearMonthクラスの`IComparable<YearMonth>.CompareTo(YearMonth)` メソッドが実行されます。

int や DateTime を OrderBy で指定してソートできる理由は、それらの型に `IComparable<T>` が実装されているからです。
![image](https://github.com/kurosu-being/TextC/assets/78195392/a1cf5a87-5e7f-4d63-bc4d-a504a711133d)
![image](https://github.com/kurosu-being/TextC/assets/78195392/5ab3eb7a-b6de-4f0f-b1b3-8c063aa1e6ca)

[リファレンス](https://learn.microsoft.com/ja-jp/dotnet/api/system.datetime?view=netframework-4.7.2&devlangs=csharp&f1url=%3FappId%3DDev15IDEF1%26l%3DJA-JP%26k%3Dk(System.DateTime)%3Bk(TargetFrameworkMoniker-.NETFramework%2CVersion%253Dv4.7.2)%3Bk(DevLang-csharp)%26rd%3Dtrue) にも記載があります。
定義のセクションの実装の部分に実装しているインターフェースが記載してあります。
![image](https://github.com/kurosu-being/TextC/assets/78195392/4db2ad68-3fd2-4b29-a3d4-ca928895cc7b)

kurosu-being replied:
ご指摘ありがとうございます！
ご提案頂いたIComparable<T> インターフェースで実装してみました。
まだインターフェイスの理解に不安がある為、ご指摘頂いたとおりに実装できているか不確定です。
恐れ入りますが、ご確認よろしくお願いします🙇‍♀️

yuki-tanaka-being replied:
大丈夫です。実装方法あっています。

yuki-tanaka-being commented on 章末問題4/Practice4-1-1/YearMonth.cs:
```diff
@@ -0,0 +1,24 @@
+        /// <summary>
+        /// 年
+        /// </summary>
+        public int Year { get; private set; }
+        /// <summary>
+        /// 月
+        /// </summary>
+        public int Month { get; private set; }
```
最後の問題まで確認しましたが、セッターは不要ですね。
クラス内でも書き換えないのであれば、セッターもなくしてイミュータブルにするのがよいです。

kurosu-being replied:
ご指摘ありがとうございます！
セッターをなくしてイミュータブルに書き直しました。

yuki-tanaka-being reviewed (commented):
レビューしましたのでご確認お願いいたします。

kurosu-being committed:
#456921 [C#テキスト]4章演習問題

プロパティの設定において、セッターもなくしイミュータブルに変更しました。
ｗDay1～ｗDay５を作るのではなく、配列の中でnewをする形に変更しました。
FindFirst21YearMonth→First21YearMonthに変更しました。
Practice4-2-5をプロパティを書き換えず、呼び出し側で新たな変数を作り格納する形に変更しました。

kurosu-being committed:
#456921 [C#テキスト]4章演習問題

Practice4-2-5において、CompareToインターフェイスを用いて書き直しました。

kurosu-being committed:
#456921 [C#テキスト]4章演習問題

インデントのズレを修正しました。

k-miyahara-being commented on 章末問題4/Practice4-2-3/YearMonth.cs:
```diff
@@ -0,0 +1,60 @@
+        /// </summary>
+        /// <param name="wYearMonths">年月</param>
+        /// <returns>21世紀の年月</returns>
+        public static YearMonth First21YearMonth(IEnumerable<YearMonth> wYearMonths) {
+            return wYearMonths.FirstOrDefault(x => x.Is21stCentury);
+        }
```
引数の接頭辞が間違っています。

kurosu-being replied:
ご質問ありがとうございます！
wYearMonths→vYearMonthsに変更いたしました。

k-miyahara-being reviewed (commented):
ほとんど言うことないです。。
田中さんのレビューでたくさん学習できたと思うので、また復習しておくと良いと思います👍
私も勉強します。

kurosu-being committed:
#458736 [C#テキスト]6章演習問題

Practice6-1-1のフォルダを作成しなおしました。

kurosu-being committed:
Revert "#458736 [C#テキスト]6章演習問題"

This reverts commit dd0013c5ad23dc40ade24bdf4e3e6dabbd181e7f.

kurosu-being committed:
#456921 [C#テキスト]4章演習問題

引数の命名を変更しました（wYearMonths→vYearMonths）
ToStringのメソッドオーバーライドの記述を簡略化しました。

k-miyahara-being reviewed (approved):
OKです！
